### ISSUE TITLE
False positive in .Rbuildignore check

### ISSUE THREAD

**Issue by LiNk-NY** [Bioconductor/BiocCheck#197]
**Edit: This issue was orginally about BiocCheck false positives in general, and not about the `.Rbuildignore` check in particular. The original discussion was moved [here](https://github.com/Bioconductor/BiocCheck/issues/201) on Dec 20, 2023.**

@LiNk-NY Not really the place here to discuss this but it's actually a mix of:
- false positives (e.g. detection of examples, vignettes, or unit tests, that trigger installation of packages),
- checks that are too stringent (e.g. ERROR instead of WARNING on use of `Sys.setenv`),
- and checks that lack consensus (e.g. enforcing use of `Authors@`).

_Originally posted by @hpages in https://github.com/Bioconductor/Rsamtools/issues/48#issuecomment-1707085694_

**Comment by LiNk-NY** [comment]
To follow up, there are a few unit tests that test for first bullet point that you make: 

Checking for install calls in R files:

https://github.com/Bioconductor/BiocCheck/blob/e87ca7a31714fed6555d980a21eb5caf3adf46cb/inst/unitTests/test_BiocCheck.R#L804-L812

and in vignettes 

https://github.com/Bioconductor/BiocCheck/blob/e87ca7a31714fed6555d980a21eb5caf3adf46cb/inst/unitTests/test_BiocCheck.R#L834-L844

For reference, the value of `.BAD_INSTALL_CALLS` is : 

```r
> BiocCheck:::.BAD_INSTALL_CALLS
[1] "biocLite"         "install.packages" "install_packages" "update.packages" 
[5] "install"      
```


Can you show where you see the false positives? Currently, these functions that check for function symbols from the parsed code are flagging installation calls in R files and _evaluated_ code in the vignettes. 

For bullet point two, it would require a wider discussion with the rest of the team. It doesn't seem like a good idea to set environment variables for the user and it is best to avoid inadvertently overwriting user environment variables.

Checks that lack consensus are usually relegated to NOTEs but a discussion is also welcome to achieve consensus.

**Comment by hpages** [comment]
There are some legit use cases for having install calls in R files. These are fine as long as they are not executed during `R CMD build` or `R CMD check`. See https://community-bioc.slack.com/archives/C0207H5QC3X/p1660313871200209 for a more in-depth discussion about this.

I agree that, generally speaking, use of `Sys.setenv` is a little bit suspicious. Unfortunately it's often a necessary evil.

For things like using `Author`/`Maintainer` instead of `Authors@`, it's really a matter of taste. Why would using the former be considered an ERROR?

These things would not be such a big deal if they were made WARNINGs or NOTEs instead of ERRORs. But the fact that they are ERRORs make it a show-stopper to add BiocCheck to the daily builds.

**Comment by LiNk-NY** [comment]
I don't think the exception should be the rule. It is better to filter out packages that do have installation calls in their R code because most often they will evaluate (from my impression). I suppose we can make an exception for `ExperimentHub` / `AnnotationHub` packages in `BiocCheck` but it is not ideal. 

`Sys.setenv` can be a bit dangerous so it's better to avoid it. One can also try using the `callr` package with environment variables set but it may be too much overhead. 

`Authors@R` is prefereable because it allows maintainers to include ORCiD IDs among other features. 


> These things would not be such a big deal if they were made WARNINGs or NOTEs instead of ERRORs. But the fact that they are ERRORs make it a show-stopper to add BiocCheck to the daily builds.

I think we have to review some of these decisions as a group.

**Comment by hpages** [comment]
> I think we have to review some of these decisions as a group.

Absolutely, but I would argue that we should maybe take the soft approach by default. That is, unless a BiocCheck test has been discussed as a group and a consensus has been reached, it should remain a WARNING. Same thing if it's known to produce false positives.

**Comment by mjsteinbaugh** [comment]
@LiNk-NY Here's a false positive I'm hitting:

```
* Checking .Rbuildignore...
    * ERROR: .Rbuildignore file includes the 'tests' folder.
```

`.Rbuildignore` contents:

```
^tests/testthat/cache$
```

I use the `tests/testthat/cache` folder to avoid re-downloading large files on a local machine, but they will be downloaded automatically during `R CMD check` on a remote machine.

I suppose alternatively I could use a `.cache` folder in the package instead...worth a thought about that.

EDIT: Alternate approach that caches test data outside of the tests folder:
https://github.com/acidgenomics/r-pipette/blob/develop/tests/testthat/helper-cache.R#L5

```
cacheDir <- file.path(
    tools::R_user_dir(package = .pkgName, which = "cache"),
    "testthat"
)
dir.create(cacheDir, showWarnings = FALSE, recursive = TRUE)
```

**Comment by mjsteinbaugh** [comment]
Also, adding `^longtests$` causes the 'tests' error above to return

**Comment by hpages** [comment]
> Also, adding `^longtests$` causes the 'tests' error above to return

Probably because `grepl("\\^*tests\\$*", entries)` is used to detect the faulty lines. Maybe it should use `"\\^tests\\$"` instead (i.e. drop the stars `*`), or `grepl("^tests$", entries, fixed=TRUE)`.

@mjsteinbaugh But why would you add `^longtests$` to your `.Rbuildignore` file? This will break the [automated long test checks](https://bioconductor.org/checkResults/devel/bioc-longtests-LATEST/).

**Update:** So I think that both `^tests$` and `^longtests$` should be forbidden.

@LiNk-NY Should we consider the risk that people could prepend slashes and dots to the paths they put in their `.Rbuildignore` file? e.g. `^longtests/$`. Maybe this discussion diserves an issue on its own.

**Comment by hpages** [comment]
> Should we consider the risk that people could prepend slashes etc...

Forget that. Because of the way `R CMD build` treats the `.Rbuildignore` file, using trailing slahes like in `^tests/$` is boggus i.e. it does not lead to the exclusion of the `tests` folder. Sorry for the noise.

**Comment by mjsteinbaugh** [comment]
@hpages I took `^longtests/$` out of my `.Rbuildignore` files but just noticed that the current error message is a bit unclear -- it says that `tests` aren't allowed when really the culprit could be `longtests`, so that could be improved a bit.

**Comment by hpages** [comment]
Agreed that the error message is confusing/misleading.

**Comment by LiNk-NY** [comment]
> > Should we consider the risk that people could prepend slashes etc...
> 
> Forget that. Because of the way `R CMD build` treats the `.Rbuildignore` file, using trailing slahes like in `^tests/$` is boggus i.e. it does not lead to the exclusion of the `tests` folder. Sorry for the noise.

@hpages My testing shows (`r85547`) that it will exclude the `tests` folder if it is added to `.Rbuildignore` as either `^tests/$`, `tests`, `tests/`, etc.
I made the update in `BiocCheck` version `1.39.1`.

**Comment by hpages** [comment]
The regular expression is too loose IMO:
```
> grepl("[\\^long]*tests[\\$\\/]*$", "^ABCDtests$")
[1] TRUE
```
It should only allow `^tests$` or  `^longtests$`. Note that you don't even need a regular expression for that.

Thanks

**Comment by LiNk-NY** [comment]
As above, a simple `^tests$` match will not capture other possibilities like `tests/` in `.Rbuildignore`. 
Folders like `.*tests` (other than `longtests`) are highly unlikely but in any case, I've updated the regex to match folders that start with `^[\\^long]...`.

**Comment by hpages** [comment]
> As above, a simple ^tests$ match will not capture other possibilities like tests/ in .Rbuildignore.

That is not needed. An entry like `^tests/$` in `.Rbuildignore` is boggus and has not effect, so you actually don't need to try to catch it.

The reg expr will still produce false positives:
```
> grepl("^[\\^long]*tests[\\$\\/]*$", "nnntests")
[1] TRUE
```
I don't think that something like `[long]*` does what you think it does. You probably want to use something like `(long)?` instead.

Thanks

**Comment by LiNk-NY** [comment]
The `^tests/$` in the previous comment was a typo. 

Currently the unit tests covers `"tests", "tests/", "^tests$", "^tests/"` : 
https://github.com/Bioconductor/BiocCheck/blob/ee9557222a7953e762cf085826e0226ab6d40feb/inst/unitTests/test_BiocCheck.R#L299-L315

Thanks for the grouping suggestion. I have updated the regex to: 

```r
grepl("^\\^?(long)?tests[\\$\\/]?$", entries, perl = TRUE)
```

and added the bogus `^tests/$` which should not be matched as you suggest.

**Comment by hpages** [comment]
Thanks Marcel.

2 last minor things, mostly as a FYI:
- Not sure why you switched to `perl=TRUE` in your last iteration. _Not_ using `perl=TRUE` actually makes things simpler since you don't need to escape `$` and `/` in `[\\$\\/]`. So why not just do `[$/]` and drop `perl=TRUE`?
- The testing strategy where you only count the number of `TRUE`s in the result seems a little loose. Sounds like it wouldn't be too hard to break the regular expression without breaking this test?

Feel free to ignore.

I'm going to rename this issue to "False positive in .Rbuildignore check" and close it.

I'll also try to move the more general discussion about BiocCheck false positives to a new issue.

**Comment by LiNk-NY** [comment]
It seems like special characters do not need to be escaped within the brackets `[ ... ]` so `[$/]` can work in this case but would not make a difference. 

> * Not sure why you switched to `perl=TRUE` in your last iteration.

My thought was that using `?` was a perl feature but I was confusing it with lookbehind syntax and it is not needed in this case. 

Thanks
